### ISSUE TITLE
octave_kernel not working on Win10

### ISSUE THREAD

**Issue by Vancikv** [Calysto/octave_kernel#45]
Hello, 

I am trying to get jupyter running on Windows 10 with an Octave kernel. I installed the octave_kernel package using pip. I get the following error message when I run `jupyter console --kernel octave` in the command prompt:

```
[IPKernelApp] ERROR | Can't load 'c:\users\kapsak\appdata\local\enthought\canopy
\user\lib\site-packages\metakernel\magics\shell_magic.py': error: <metakernel.pe
xpect.spawnu object at 0x00000000045E4048>
version: 3.3
command: C:\Windows\system32\cmd.EXE
args: ['C:\\Windows\\system32\\cmd.EXE']
searcher: None
buffer (last 100 chars): u'Microsoft Windows [Version 10.0.10586]'
before (last 100 chars): u'Microsoft Windows [Version 10.0.10586]'
after: <class 'metakernel.pexpect.TIMEOUT'>
match: None
match_index: None
exitstatus: None
flag_eof: False
pid: 2804
child_fd: 4
closed: False
timeout: 30
delimiter: <class 'metakernel.pexpect.EOF'>
logfile: None
logfile_read: None
logfile_send: None
maxread: 2000
ignorecase: False
searchwindowsize: None
delaybeforesend: 0.05
delayafterclose: 0.1
delayafterterminate: 0.1
version: 3.3
command: C:\Windows\system32\cmd.EXE
args: ['C:\\Windows\\system32\\cmd.EXE']
searcher: None
buffer (last 100 chars): u'Microsoft Windows [Version 10.0.10586]'
before (last 100 chars): u'Microsoft Windows [Version 10.0.10586]'
after: <class 'metakernel.pexpect.TIMEOUT'>
match: None
match_index: None
exitstatus: None
flag_eof: False
pid: 2804
child_fd: 4
closed: False
timeout: 30
delimiter: <class 'metakernel.pexpect.EOF'>
logfile: None
logfile_read: None
logfile_send: None
maxread: 2000
ignorecase: False
searchwindowsize: None
delaybeforesend: 0.05
delayafterclose: 0.1
delayafterterminate: 0.1
```

I'll be grateful for any hints, we want to have students use Jupyter notebooks with Octave, so I need to be sure I can get it running on Windows.

**Comment by blink1073** [comment]
Hi @Vancikv, I just verified that I can run the Octave Kernel on Windows 10 using Octave 4.0.3.

The steps I used were:

```bash
conda create -n test jupyter
source activate test
pip install octave_kernel
python -m octave_kernel.install
jupyter console --kernel octave
```

@dsblank, have you seen any problems with the shell magic on Windows?

**Comment by dsblank** [comment]
I must admit that I have never used any Jupyter-server-related stuff on Windows. (If I were to, I'd probably use that new ubuntu virtual machine thing.)

**Comment by Vancikv** [comment]
Thanks for the feedback, I used Anaconda and was only successful with Python 3 although I have no trouble running octave_kernel with Python 2 on Ubuntu. So I have managed to get it running but some problems still persist:

1) Running the octave console works but when I deliberately run an invalid statement (like use an undefined variable as RHS of equation) I get a similar error message as I posted before. This happens also when I run the "quit" command, so I can't close the console properly.

2) When I run a notebook that uses octave kernel and run an octave code cell out of turn (before another cell that it depends on), I get the following error output:

`[IPKernelApp] ERROR | Exception in message handler:
Traceback (most recent call last):
  File "C:\Users\Werner\Anaconda3\lib\site-packages\metakernel\pexpect.py", line
 1251, in _send
    return os.write(self.write_fd, s)
OSError: [Errno 22] Invalid argument

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\Werner\Anaconda3\lib\site-packages\metakernel\process_metakerne
l.py", line 80, in do_execute_direct
    stream_handler=stream_handler)
  File "C:\Users\Werner\Anaconda3\lib\site-packages\metakernel\replwrap.py", lin
e 138, in run_command
    send_prompt_emit=val != 2)
  File "C:\Users\Werner\Anaconda3\lib\site-packages\metakernel\replwrap.py", lin
e 97, in _expect_prompt
    self.sendline(self.prompt_emit_cmd)
  File "C:\Users\Werner\Anaconda3\lib\site-packages\metakernel\replwrap.py", lin
e 87, in sendline
    self.child.sendline(line)
  File "C:\Users\Werner\Anaconda3\lib\site-packages\metakernel\pexpect.py", line
 1263, in sendline
    n = n + self.send(self.linesep)
  File "C:\Users\Werner\Anaconda3\lib\site-packages\metakernel\pexpect.py", line
 1242, in send
    return self._send(s)
  File "C:\Users\Werner\Anaconda3\lib\site-packages\metakernel\pexpect.py", line
 1945, in _send
    return super(spawnu, self)._send(s.encode(self.encoding, self.errors))
  File "C:\Users\Werner\Anaconda3\lib\site-packages\metakernel\pexpect.py", line
 1254, in _send
    raise EOF('Attempted write to a closed file descriptor')
metakernel.pexpect.EOF: Attempted write to a closed file descriptor

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\Werner\Anaconda3\lib\site-packages\ipykernel\kernelbase.py", li
ne 228, in dispatch_shell
    handler(stream, idents, msg)
  File "C:\Users\Werner\Anaconda3\lib\site-packages\ipykernel\kernelbase.py", li
ne 391, in execute_request
    user_expressions, allow_stdin)
  File "C:\Users\Werner\Anaconda3\lib\site-packages\metakernel\_metakernel.py",
line 362, in do_execute
    retval = self.do_execute_direct(code)
  File "C:\Users\Werner\Anaconda3\lib\site-packages\octave_kernel\kernel.py", li
ne 104, in do_execute_direct
    super(OctaveKernel, self).do_execute_direct(code, self.Print)
  File "C:\Users\Werner\Anaconda3\lib\site-packages\metakernel\process_metakerne
l.py", line 90, in do_execute_direct
    self._start()
  File "C:\Users\Werner\Anaconda3\lib\site-packages\metakernel\process_metakerne
l.py", line 57, in _start
    self.wrapper.child.terminate()
  File "C:\Users\Werner\Anaconda3\lib\site-packages\metakernel\pexpect.py", line
 1332, in terminate
    self.kill(signal.SIGTERM)
  File "C:\Users\Werner\Anaconda3\lib\site-packages\metakernel\pexpect.py", line
 1504, in kill
    os.kill(self.pid, sig)
PermissionError: [WinError 5] Access is denied`

3) When I run a cell containing an octave function, I get: `warning: function name 'name_of_the_function' does not agree with function file name ''` which then results in 2) when I run a cell that wants to use that function.

Anyway, it turns out we'll just be using Jupyter for teaching and students will only work with octave itself, so If you don't want to spend more time on this, you need not do so for my sake.

**Comment by blink1073** [comment]
Supporting Windows has been very tricky, because it does not have support for the `pty` module, which makes things seamless on Posix systems.  For example, a syntax error on Windows causes Octave to forcibly close the pipe we have opened, leaving us with no way to recover the session.

**Comment by blink1073** [comment]
This is fixed in v0.19, thank you again for the report.
